src/commands/signzone.rs

```diff
@@ -0,0 +1,4157 @@
+            //     private_key_path.display()
+            // );
+        }
+
+        // First split the key into Key Signing Keys (KSK) that sign the
+        // DNSKEY RRset and Zone Signing Keys (ZSK) that sign the zone.
+        let mut key_signing_keys = Vec::new();
+        let mut zone_signing_keys = Vec::new();
+        for k in &signing_keys {
+            if k.is_secure_entry_point() {
+                key_signing_keys.push(k);
+            } else {
+                zone_signing_keys.push(k);
+            }
+        }
+
+        if key_signing_keys.is_empty() {
+            // Sign the DNSKEY RRset with the zone signing keys.
+            key_signing_keys.append(&mut zone_signing_keys.clone());
+        } else if zone_signing_keys.is_empty() {
+            // Sign the zone with the key signing keys.
+            zone_signing_keys.append(&mut key_signing_keys.clone());
+        } else {
+            if self.sign_dnskeys_with_all_keys {
+                // Sign DNSKEY RRset with all keys. Add the ZSKs to the
+                // KSKs.
+                key_signing_keys.append(&mut zone_signing_keys.clone());
+            }
+            if self.sign_with_every_unique_algorithm {
+                // Add ZSKs to KSKs if the ZSKs have an algorithm that is
+                // not currently used by the KSKs.
+                let mut algorithms = HashSet::new();
+                for k in &key_signing_keys {
+                    algorithms.insert(k.algorithm());
+                }
+                for k in &zone_signing_keys {
+                    if !algorithms.contains(&k.algorithm()) {
+                        // ldns-signzone adds just one key per algorithm.
+                        algorithms.insert(k.algorithm());
+
+                        key_signing_keys.push(k);
+                    }
+                }
+
+                // Add KSKs to ZSKs if the KSKs have an algorithm that is
+                // not currently used by the ZSKs.
+                let mut algorithms = HashSet::new();
+                for k in &zone_signing_keys {
+                    algorithms.insert(k.algorithm());
+                }
+                for k in &key_signing_keys {
+                    if !algorithms.contains(&k.algorithm()) {
+                        // ldns-signzone adds just one key per algorithm.
+                        algorithms.insert(k.algorithm());
+
+                        zone_signing_keys.push(k);
+                    }
+                }
+            }
+        }
+
+        let mut dnskey_extra = Vec::new();
+        let mut all_dnskeys = Vec::new();
+        let empty_records: [Record<_, _>; 0] = [];
+        for r in dnskey_rrset
+            .as_ref()
+            .map_or(empty_records.iter(), |r| r.iter())
+        {
+            all_dnskeys.push(r.clone());
+        }
+        if !self.do_not_add_keys_to_zone {
+            let dnskey_ttl = dnskey_rrset.as_ref().map_or(soa_rr.ttl(), |r| r.ttl());
```

Comment:
The [original code](https://github.com/NLnetLabs/domain/pull/416/files/fcc94d299cb0cfadb88c388aec7aabaa2da12d5b..9e166c64bbac3893eac2b4b763bea4b28b2c7f91#diff-d1299cb884bc24469f78e37c88ce3432bd5ae47f427d6eebaf34bfabe74601f1L477) in domain that used to do this had a comment explaining why it used the SOA TTL as a fallback.

I think it would be good to still document this.

Perhaps add the following text? Though ideally I'd prefer to have an RFC reference to point to if you know of one that applies here...

```suggestion
            // Use the TTL of existing DNSKEY RRs for any DNSKEY RRs that we
            // add (as all RRs in an RRSET MUST have the same TTL) else use
            // the the SOA RR TTL when there are no existing DNSKEY RRs. (as
            // that is what dnssec-signzone and ldns-signzone do).
            let dnskey_ttl = dnskey_rrset.as_ref().map_or(soa_rr.ttl(), |r| r.ttl());
```